### PR TITLE
Long tap on route to lock value sent from this network

### LINKED ISSUE #16974: Implement Send page > Long tap to edit amount send through an specific network
# Feature Issue

### User Story

As a user, I want to long tap to edit the amount of tokens sent through a specific network so that I can customize the transaction value for that network.

### Description

Figma Design 1: https://www.figma.com/file/HKncH4wnDwZDAhc4AryK8Y/Wallet-for-Mobile?type=design&node-id=2388-503644&mode=design&t=V0dXiJUCqM0Jo7Od-4
Figma Design 2: https://www.figma.com/file/HKncH4wnDwZDAhc4AryK8Y/Wallet-for-Mobile?type=design&node-id=2388-503816&mode=design&t=V0dXiJUCqM0Jo7Od-4
Figma Design 3: https://www.figma.com/file/HKncH4wnDwZDAhc4AryK8Y/Wallet-for-Mobile?type=design&node-id=2436-519277&mode=design&t=V0dXiJUCqM0Jo7Od-4
Figma Design 4: https://www.figma.com/file/HKncH4wnDwZDAhc4AryK8Y/Wallet-for-Mobile?type=design&node-id=2436-519599&mode=design&t=V0dXiJUCqM0Jo7Od-4
Figma Design 5: https://www.figma.com/file/HKncH4wnDwZDAhc4AryK8Y/Wallet-for-Mobile?type=design&node-id=2436-519846&mode=design&t=V0dXiJUCqM0Jo7Od-4

Visual reference:
    

### Acceptance Criteria

- [x] When user long-taps network card in the left column, `Send from <network>` bottom sheet should be shown 
- [x] User can edit the amount sent through that network
- [x] User can press the button at the right of the input to toggle the value in its preferred currency
- [x] When user edit the amount, checkbox with label `Don’t auto recalculate <network>` should be automatically checked
- [x] Whenever the user wants the amount be auto recalculated for that network, the checkbox should be manually unchecked
- [x] User should be allowed to input any values in this drawer but if they exceed available maximum on that network we'll have an error like on the picture 
- [ ] If user inputs less than maximum, but other networks do not have enough to cover required amount I believe it should be something like on the picture  (Check with design team)
- [x] User should tap the `Update` button to apply changes.

Feel free to break up this issue in smaller issues if necessary

### PR DESCRIPTION
**Pull request by vkjr**
fixes #16974

### Summary

Implemented long tap to edit amount of tokens sent through the specific network

https://github.com/status-im/status-mobile/assets/5786310/84a41136-8613-42ae-b217-20507a62195e

### Review notes
Implemented screen borrows heavily from `send/input_amount/view.cljs`, mostly because of `token-input` component. So there is definitely room for a refactoring but here I'd like to concentrate more on functionality. And postpone refactoring until we decide to refactor `send/input_amount/view.cljs`

##### Functional

- wallet / transactions

status: ready

### CODE REVIEW

**Comment by ulisesmac** [comment]
@vkjr Just a question about the video, why you are able to set `1.1` in the modal when you only set `1` in the screen behind? Is it expected or it still lacks of validation? if so, do we have an issue for it?

File: src/status_im/contexts/wallet/send/events.cljs [hunk 1]
```diff
@@ -37,6 +37,7 @@
            sender-network-values         (get-in db [:wallet :ui :send :sender-network-values])
            tx-type                       (get-in db [:wallet :ui :send :tx-type])
            disabled-from-chain-ids       (or (get-in db [:wallet :ui :send :disabled-from-chain-ids]) [])
+           from-locked-amounts           (or (get-in db [:wallet :ui :send :from-locked-amounts]) {})
```
Comment (ulisesmac) [inline src/status_im/contexts/wallet/send/events.cljs]: `get-in` already has a `not-found` version, better to use it than using `or`

File: src/status_im/contexts/wallet/send/utils.cljs [hunk 2]
```diff
@@ -156,8 +157,12 @@
               (map
                (fn [[chain-id amount]]
                  {:chain-id     chain-id
-                  :total-amount amount
+                  :total-amount (let [locked-amount (get from-locked-amounts chain-id)]
+                                  (if locked-amount
+                                    locked-amount
+                                    amount))
```
Comment (ulisesmac) [inline src/status_im/contexts/wallet/send/utils.cljs]: We can refactor the whole `let` to:
`(get from-locked-amounts chain-id amount)`

File: src/status_im/subs/wallet/networks.cljs [hunk 3]
```diff
@@ -80,6 +80,7 @@
  (fn [networks [_ chain-id]]
    (some #(when (= chain-id (:chain-id %)) %) networks)))
 
+
```
Comment (ulisesmac) [inline src/status_im/subs/wallet/networks.cljs]: Extra space, worths removing it because this is the only change in this file :+1:

File: src/status_im/contexts/wallet/send/input_amount/view.cljs [hunk 4]
```diff
@@ -264,7 +264,32 @@
...
+                                                                           input-state)
+                                                                new-value (if swap-to-crypto-currency?
+                                                                            (.toFixed (/ value
+                                                                                         conversion-rate)
+                                                                                      crypto-decimals)
+                                                                            (.toFixed (* value
+                                                                                         conversion-rate)
+                                                                                      12))]
+                                                            (number/remove-trailing-zeroes
+                                                             new-value))))))]
```
Comment (ulisesmac) [inline src/status_im/contexts/wallet/send/input_amount/view.cljs:317]: Maybe we can move this to a separate function so that it has a name and it'll be easier to understand what this is doing, also the code here will look better, wdyt?

File: src/status_im/contexts/wallet/send/events.cljs [hunk 5]
```diff
@@ -308,6 +314,17 @@
...
+ (fn [{:keys [db]} [chain-id amount]]
+   {:db (assoc-in db [:wallet :ui :send :from-locked-amounts chain-id] amount)}))
+
+(rf/reg-event-fx :wallet/unlock-from-amount
+ (fn [{:keys [db]} [chain-id]]
+   (let [new-locked-amounts (-> db
+                                (get-in [:wallet :ui :send :from-locked-amounts])
+                                (dissoc chain-id))]
+     {:db (assoc-in db [:wallet :ui :send :from-locked-amounts] new-locked-amounts)})))
+
```
Comment (ulisesmac) [inline src/status_im/contexts/wallet/send/events.cljs]: We actually want to update the entry, so `update-in` could make this code easier to grasp:

```clojure
{:db (update-in db [:wallet :ui :send :from-locked-amounts] dissoc chain-id)}
```

wdyt?

**Review by ulisesmac — COMMENTED** [review]
Hey @vkjr 

Thanks for your PR!
I left some comments most related to code

File: src/status_im/contexts/wallet/send/input_amount/view.cljs (hunk 4 above)
Comment (OmarBasem) [reply src/status_im/contexts/wallet/send/input_amount/view.cljs:317]: We need to stop adding new function definitions inside that let block of the send screen, it is already too big

File: src/status_im/contexts/wallet/send/routes/view.cljs [hunk 6]
```diff
@@ -89,31 +85,171 @@
...
+             crypto-decimals                            token-decimals
+             input-amount                               (controlled-input/input-value input-state)
+             [is-amount-locked? set-is-amount-locked]   (rn/use-state (some? locked-amount))
+             bottom                                     (safe-area/get-bottom)
+             amount-in-crypto                           (if crypto-currency?
+                                                          input-amount
+                                                          (number/remove-trailing-zeroes
+                                                           (.toFixed (/ input-amount
+                                                                        conversion-rate)
+                                                                     crypto-decimals)))
```
Comment (OmarBasem) [inline src/status_im/contexts/wallet/send/routes/view.cljs:135]: I see some duplication in this block similar to the `input_amount` screen. Potentially a place for bugs in the future if changes were not synced. I think we need to make it more DRY and reusable, for example common subs and utils can be grouped into a common `init` method, for example: https://github.com/status-im/status-mobile/blob/develop/src/status_im/contexts/chat/messenger/composer/utils.cljs#L187

Comment (vkjr) [reply src/status_im/contexts/wallet/send/routes/view.cljs:135]: @OmarBasem, yes there is a duplication as I stated in PR description. And I desperately want to remove it and also to refactor `input_amount` because it is terrifying) But I'd be happy to do this after this PR merged, just to have the functionality in place.

File: src/status_im/contexts/wallet/send/input_amount/view.cljs (hunk 4 above)
Comment (vkjr) [reply src/status_im/contexts/wallet/send/input_amount/view.cljs:317]: Well, I totally agree with you but this is not a new function, I just moved handlers to `let` section, so the code down below is easier to comprehend: `:on-swap swap-between-fiat-and-crypto`.
But this let block will benefit from reduction, for sure. I will try to move some functions outside

File: src/status_im/contexts/wallet/send/events.cljs (hunk 5 above)
Comment (vkjr) [reply src/status_im/contexts/wallet/send/events.cljs]: Yes, this is much better, thank you!

File: src/status_im/contexts/wallet/send/events.cljs (hunk 1 above)
Comment (vkjr) [reply src/status_im/contexts/wallet/send/events.cljs]: Great, thanks!

File: src/status_im/contexts/wallet/send/utils.cljs (hunk 2 above)
Comment (vkjr) [reply src/status_im/contexts/wallet/send/utils.cljs]: Thanks a lot! I need to memorize optional arguments of common functions :))

**Comment by vkjr** [comment]
> @vkjr Just a question about the video, why you are able to set `1.1` in the modal when you only set `1` in the screen behind? Is it expected or it still lacks of validation? if so, do we have an issue for it?

Well, technically I don't think there should be any issue with locking amount which is higher than currently typed value, because we can change total value after we changed locked one. If we force the user to type total value first and locks only after this that wouldn't be the best UX (my opinion)
But here I put the value higher because when I put equal value I'm getting "no routes found" and I didn't investigate this issue yet.
And when I put value lower I'm also getting "no routes found" and other networks not taken into account, but this is a problem on router side, because desktops behavior was the same.

Overall there are few bugs and questions to the designers, I will create additional issues.

**Comment by vkjr** [comment]
@ulisesmac , @OmarBasem , I've fixed minor notes with the goal to make bigger refactorings later and separately

**Review by ulisesmac — APPROVED** [review]
@vkjr 
Thanks for addressing the comments! :+1:

File: src/status_im/contexts/wallet/common/utils.cljs [hunk 7]
```diff
@@ -269,3 +269,15 @@
     (some #(when (= token-symbol (:symbol %))
              %)
           address-tokens)))
+
```
Comment (J-Son89) [inline src/status_im/contexts/wallet/common/utils.cljs:272]: mmhh, can this go into some more categorised utils?

we have common/utils/<some-category>

this file is already a bit of a mess of random functionality :)

File: src/status_im/contexts/wallet/send/events.cljs [hunk 8]
```diff
@@ -348,7 +363,8 @@
          to-token-id ""
          network-preferences (if token [] [(get-in collectible [:id :contract-id :chain-id])])
          gas-rates constants/gas-rate-medium
-         amount-in (send-utils/amount-in-hex amount (if token token-decimal 0))
+         to-hex (fn [v] (send-utils/amount-in-hex v (if token token-decimal 0)))
+         amount-in (to-hex amount)
```
Comment (J-Son89) [inline src/status_im/contexts/wallet/send/events.cljs:390]: why did you need to do this?

File: src/status_im/contexts/wallet/send/input_amount/view.cljs (hunk 4 above)
Comment (J-Son89) [reply src/status_im/contexts/wallet/send/input_amount/view.cljs:317]: haha, yeah let's start to split this file up somehow 😅

**Review by J-Son89 — APPROVED** [review]

**Review by OmarBasem — APPROVED** [review]

File: src/status_im/contexts/wallet/send/events.cljs (hunk 8 above)
Comment (vkjr) [reply src/status_im/contexts/wallet/send/events.cljs:390]: Because `to-hex` is used one more time below)

**Comment by vkjr** [comment]
@VolodLytvynenko, there was an issue with rebase, I've fixed it

**Comment by vkjr** [comment]
After discussions some changes should be added to this feature:
https://www.figma.com/design/xLs1KYmF4e6WwRTZVJKeUK/Descoped-Wallet?node-id=15901-45021&t=fPVa2luan0Vf3DZO-1

https://www.figma.com/design/xLs1KYmF4e6WwRTZVJKeUK/Descoped-Wallet?node-id=15902-46799&t=OAFVxkZt9a5YyxQE-1

@VolodLytvynenko, while we are waiting for design to agree with John on these decisions, wdyt about testing/fixing/merging PR as is and adding the aforementioned designs in separate PRs?

**Comment by VolodLytvynenko** [comment]
> @VolodLytvynenko, while we are waiting for design to agree with John on these decisions, wdyt about testing/fixing/merging PR as is and adding the aforementioned designs in separate PRs?

@vkjr Yes, I fully agree to add those features separately. For now, the PR is in the testing progress.
Just an update: the PR looks good, but it took me some time to understand how it works and where the related issues are. I found some issues related to Status Go which are logged separatly:

https://github.com/status-im/status-mobile/issues/20486
https://github.com/status-im/status-mobile/issues/20484

And some low-priority issues related to the current PR. Today I will finish testing

**Comment by vkjr** [comment]
@VolodLytvynenko, thanks a lot!

**Comment by VolodLytvynenko** [comment]
Can be considered as follow-up:

## PR_ISSUE 1: The initial value is not predefined in the drawer after network is long-pressed

### Steps:
1. Go to the routes generation page
2. Enter the value that is available on at least one network
3. Long press on the network when the route is built

### Actual result:
The value is not predefined in the drawer after the network is long-pressed.

https://github.com/status-im/status-mobile/assets/52490791/e443599e-53ca-4a21-b766-6dcb8d72e2ea


### Expected result:
The value is predefined in the drawer after the network is long-pressed.


### OS: 
IOS, Android
### Devices:
- Pixel 7a, Android 13
- iPhone 11 Pro Max, IOS 17

**Comment by VolodLytvynenko** [comment]
Can be considered as follow-up:

## PR_ISSUE 2:  The long-press is available for disabled network

### Steps:
1. Go to the routes generation page
2. Enter the value that is available on at least two networks
3. Disable the network that is built (just short tap on it)
4. Long press on the disabled network


### Actual result:
The drawer is opened

https://github.com/status-im/status-mobile/assets/52490791/b54aa71e-44a6-44d0-969e-dcafff1777a0

### Expected result:
The drawer is not opened when the disabled network is long-pressed

### Devices:
- Pixel 7a, Android 13
- iPhone 11 Pro Max, IOS 17

**Comment by VolodLytvynenko** [comment]
Can be considered as follow-up:

## PR_ISSUE 3: The "Not enough assets to pay gas fees" caution is not shown within the long-pressed drawer when max value is entered

### Steps:
1. Go to the routes generation page
2. Enter the value that is available on at least one network
3. Long press on the network when the route is built
4. Enter the max value

### Actual result:
The "Not enough assets to pay gas fees" caution is not shown

### Expected result:
The "Not enough assets to pay gas fees" caution is shown as implemented within the routes generation page??? 
(need to be discussed with design team)


### OS: 
IOS, Android
### Devices:
- Pixel 7a, Android 13
- iPhone 11 Pro Max, IOS 17

**Comment by vkjr** [comment]
@VolodLytvynenko, if you already have finished testing and these are all found issues, I'd be glad to fix them in a separate PR as followup, wdyt?

**Comment by VolodLytvynenko** [comment]
## PR_ISSUE 4: No ability to distribute amount for other networks if 'don't auto recalculate <network>' is unchecked

### Steps:
1. Go to the routes generation page.
2. Enter a value that is available on at least 2 networks (e.g., 0.003; the arb route is built).
3. Long press on the network where the route is not built yet (e.g., eth mainnet).
4. Enter an available amount in the 'Define amount' drawer (e.g., 0.001).
5. Uncheck the 'don't auto recalculate <network>' checkbox.

### Actual result:
The amount is not distributed to the recently selected network if 'don't auto recalculate <network>' is unchecked.

https://github.com/status-im/status-mobile/assets/52490791/3bf2b97c-fd11-4ae5-8b81-871459815025

### Expected result:
The amount is distributed to the recently selected network if 'don't auto recalculate <network>' is unchecked in the following way:
- 0.002 arb -> arb
- 0.001 mainnet -> mainnet

### Devices:
- Pixel 7a, Android 13
- iPhone 11 Pro Max, IOS 17

**Comment by VolodLytvynenko** [comment]
Hi @vkjr PR can be merged. All issues can be fixed separately. Thank you

**Comment by vkjr** [comment]
@VolodLytvynenko, regarding issue 4:
During implementation my understanding was that the checkbox "don't auto recalculate network" is an indicator that shows if amount is locked. This is why it is automatically checked when you change the value. But when you uncheck it, that means that you allow the router to automatically recalculate network amount. And possible scenario is that router won't use this network.
So I treat this checkbox as the only way to unlock the previously locked amount.
In the [original issue](https://github.com/status-im/status-mobile/issues/16974) there is a subtask saying "Whenever the user wants the amount to be auto recalculated for that network, the checkbox should be manually unchecked"

But maybe I'm incorrectly perceiving the purpose of the checkbox, wdyt?

**Comment by vkjr** [comment]
Followup issues created:

https://github.com/status-im/status-mobile/issues/20498
https://github.com/status-im/status-mobile/issues/20497
https://github.com/status-im/status-mobile/issues/20496

cc @VolodLytvynenko

**Comment by VolodLytvynenko** [comment]
> @VolodLytvynenko, regarding issue 4: During implementation my understanding was that the checkbox "don't auto recalculate network" is an indicator that shows if amount is locked. This is why it is automatically checked when you change the value. But when you uncheck it, that means that you allow the router to automatically recalculate network amount. And possible scenario is that router won't use this network. So I treat this checkbox as the only way to unlock the previously locked amount. In the [original issue](https://github.com/status-im/status-mobile/issues/16974) there is a subtask saying "Whenever the user wants the amount to be auto recalculated for that network, the checkbox should be manually unchecked"
> 
> But maybe I'm incorrectly perceiving the purpose of the checkbox, wdyt?

@vkjr Yes. It makes sense. Now I've rethought it. I'll try to find the correct answer :)

**Comment by VolodLytvynenko** [comment]
one more follow-up https://github.com/status-im/status-mobile/issues/20526
